### PR TITLE
#13394: Galaxy 2cq support

### PR DESCRIPTION
**Pull request by ubcheema**
Implement 2 hardware CQs for Galaxy.

https://github.com/tenstorrent/tt-metal/issues/13394

### CODE REVIEW

**Review by aliuTT — APPROVED** [review]

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/common/command_queue_fixture.hpp [hunk 1]
```diff
@@ -26,8 +27,10 @@ class MultiCommandQueueSingleDeviceFixture : public ::testing::Test {
         arch_ = tt::get_arch_from_string(tt::test_utils::get_env_arch_name());
         DispatchCoreType dispatch_core_type = DispatchCoreType::WORKER;
         if (arch_ == tt::ARCH::WORMHOLE_B0 and tt::tt_metal::GetNumAvailableDevices() != 1) {
```
Comment (pgkeller) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch_single_chip_multi_queue/common/command_queue_fixture.hpp:29]: fix this (and other places below) for BH?

File: tt_metal/impl/dispatch/command_queue_interface.hpp [hunk 2]
```diff
@@ -92,12 +94,12 @@ struct dispatch_constants {
...
 
    private:
-    dispatch_constants(const CoreType &core_type) {
+    dispatch_constants(const CoreType &core_type, const uint32_t num_hw_cqs) {
         TT_ASSERT(core_type == CoreType::WORKER or core_type == CoreType::ETH);
         // make this 2^N as required by the packetized stages
         uint32_t dispatch_buffer_block_size;
         if (core_type == CoreType::WORKER) {
-            prefetch_q_entries_ = 1534;
+            prefetch_q_entries_ = 1532 / num_hw_cqs;
```
Comment (pgkeller) [inline tt_metal/impl/dispatch/command_queue_interface.hpp:102]: comment for 1532 magic?

File: tt_metal/impl/device/device.cpp [hunk 3]
```diff
@@ -915,99 +955,278 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
...
+                    arg_index = 26;
+                    uint32_t demux_sem = demux_settings.producer_semaphore_id;
+                    for (auto&[core, settings] : device_worker_variants[DispatchWorkerType::DISPATCH]) {
+                        // 26, 27, 28, 29: output x depacketize info:
+                        compile_args[arg_index++] = packet_switch_4B_pack(settings.cb_log_page_size,
+                                                                            settings.consumer_semaphore_id, // downstream sem
+                                                                            demux_sem++,    // local sem
+                                                                            1); // remove header
+                    }
+                } else if (device_worker_variants[DispatchWorkerType::DEMUX].size() == 3) {
```
Comment (pgkeller) [inline tt_metal/impl/device/device.cpp:1004]: wow, this dispatch init code is approaching total inscrutability

**Review by pgkeller — APPROVED** [review]
a few minor things.  the contents are extremely hard to review due to our need to clean up how dispatch init occurs.  I skimmed through and didn't see obvious issues, code is clean enough given the base code, but, yikes.

File: tt_metal/impl/device/device.cpp [hunk 4]
```diff
@@ -739,22 +739,28 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
...
-                TT_ASSERT(device_worker_variants[DispatchWorkerType::MUX].size() == 1, "Cannot have more than one Mux.");
-                auto mux_settings = std::get<1>(device_worker_variants[DispatchWorkerType::MUX][0]);
-                TT_ASSERT(num_prefetchers == mux_settings.semaphores.size(), "Mux does not have required number of semaphores for Prefetchers. Exptected = {}. Found = {}", num_prefetchers, mux_settings.semaphores.size());
-                uint32_t mux_sem = mux_settings.consumer_semaphore_id;
+                uint32_t mux_count = device_worker_variants[DispatchWorkerType::MUX].size();
+                TT_ASSERT((num_prefetchers / mux_count) <= MAX_SWITCH_FAN_IN, "Insufficient Mux cores. Expected = {}. Found = {}", num_prefetchers, mux_count);
+                //auto mux_settings = std::get<1>(device_worker_variants[DispatchWorkerType::MUX][0]);
+                //TT_ASSERT(num_prefetchers == (mux_count * mux_settings.semaphores.size()), "Mux does not have required number of semaphores for Prefetchers. Exptected = {}. Found = {}", num_prefetchers, mux_settings.semaphores.size());
+                //uint32_t mux_sem = mux_settings.consumer_semaphore_id;
```
Comment (tt-asaigal) [inline tt_metal/impl/device/device.cpp:746]: Lets get rid of this unused code

File: tt_metal/impl/device/device.cpp [hunk 5]
```diff
@@ -739,22 +739,28 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
...
+                    uint32_t downstream_cb_base = mux_settings.cb_start_address + mux_settings.cb_size_bytes * mux_sem[mux_index];
+                    uint32_t downstream_cb_pages = mux_settings.cb_pages;
                     settings.upstream_cores.push_back(tt_cxy_pair(0, 0, 0));
                     settings.downstream_cores.push_back(mux_settings.worker_physical_core);
                     settings.compile_args.resize(28);
                     auto& compile_args = settings.compile_args;
                     compile_args[0]  = downstream_cb_base;
                     compile_args[1]  = dispatch_constants::PREFETCH_D_BUFFER_LOG_PAGE_SIZE;
-                    compile_args[2]  = dispatch_constants::get(dispatch_core_type).mux_buffer_pages(num_hw_cqs);
+                    compile_args[2]  = downstream_cb_pages; //dispatch_constants::get(dispatch_core_type).mux_buffer_pages(num_hw_cqs);
```
Comment (tt-asaigal) [inline tt_metal/impl/device/device.cpp:761]: Same as above  here

**Review by ttmchiou — APPROVED** [review]

File: tt_metal/impl/device/device.cpp [hunk 6]
```diff
@@ -915,99 +955,278 @@ void Device::update_workers_build_settings(std::vector<std::vector<std::tuple<tt
...
+                        compile_args[26] = true; // is_host_variant
+
+                        dispatch_idx++;
+                    }
+                } else if (num_dispatchers == 4 || num_dispatchers == 8) {
+                    TT_ASSERT(device_worker_variants[DispatchWorkerType::DEMUX].size() == 3, "Insufficient Demux cores. Expected = 3. Found = {}", device_worker_variants[DispatchWorkerType::DEMUX].size());
+                    uint32_t dispatch_idx = 0;
+                    uint32_t demux_fanout = num_dispatchers / 2;
+                    for (int i = 1; i < 3; i++) {
+                        auto demux_settings = std::get<1>(device_worker_variants[DispatchWorkerType::DEMUX][i]);
```
Comment (tt-asaigal) [inline tt_metal/impl/device/device.cpp:1188]: can we please add a comment explaining the topology here? I've scared that debug will be very painful for people who haven't brought up this feature, if they can't understand what's going on physically.

**Review by tt-asaigal — APPROVED** [review]

File: tt_metal/impl/device/device.cpp (hunk 5 above)
Comment (ubcheema) [reply tt_metal/impl/device/device.cpp]: done

File: tt_metal/impl/device/device.cpp (hunk 4 above)
Comment (ubcheema) [reply tt_metal/impl/device/device.cpp]: done

File: tt_metal/impl/device/device.cpp (hunk 6 above)
Comment (ubcheema) [reply tt_metal/impl/device/device.cpp:1179]: I will update the tunneling diagram to show how N300 and Galaxy 1-2 CQ tunnels get connected.

Comment (tt-asaigal) [reply tt_metal/impl/device/device.cpp:1179]: Thanks!

**Pull request merged**